### PR TITLE
feat: add StripGitCredentials and NormalizeGitURL [IDE-1915]

### PR DESCRIPTION
**Pull request by bastiandoetsch**
## Summary

- Adds `StripGitCredentials()` to remove userinfo (username/password/token) from git remote URLs, preventing credential leakage to APIs and logs
- Adds `NormalizeGitURL()` to convert any git URL format (SSH, SCP-style, git://, http://) into a consistent `https://` URL with credentials removed and `.git` suffix stripped
- Handles edge cases: SSH ports stripped during normalization, `git+ssh://`/`ssh+git://` schemes, GitLab subgroups, `file://` passthrough

## Test plan

- [x] 9 table-driven test cases for `StripGitCredentials` (empty, HTTPS with user:pass, OAuth tokens, username-only, HTTP creds, SSH userinfo, SCP-style passthrough, clean URLs, git protocol)
- [x] 18 table-driven test cases for `NormalizeGitURL` (all URL formats, SSH with ports, credential stripping, `git+ssh`/`ssh+git`, GitLab subgroups, `file://` passthrough)
- [x] All 6 existing tests continue to pass
- [x] `go vet` clean

___

### **PR Type**
enhancement, bug_fix

___

### **Description**
- Add Git URL sanitization functions.

- Remove sensitive credentials from URLs.

- Normalize diverse URL formats to HTTPS.

- Include extensive unit tests.

___

<details> <summary><h3> File Walkthrough</h3></summary>

<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Enhancement</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>git.go</strong><dd><code>Implement Git URL credential stripping and normalization.</code></dd></summary>
<hr>

pkg/utils/git/git.go

<ul><li>Adds <code>StripGitCredentials</code> to remove userinfo from Git URLs.<br> <li> Implements <code>NormalizeGitURL</code> to convert various URL formats to HTTPS.<br> <li> Introduces helper functions <code>isSCPStyle</code> and <code>stripCredentialsByPattern</code>.<br> <li> Handles edge cases for malformed URLs and SCP-style syntax.</ul>

</details>

  </td>
  <td><a href="https://github.com/snyk/go-application-framework/pull/586/changes#diff-af8adb2f32344eb281e8436fafb1ea3b2b455a532b195f67cc245515f1e928a1">+134/-0</a>&nbsp; </td>

</tr>
</table></td></tr><tr><td><strong>Tests</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>git_test.go</strong><dd><code>Add comprehensive tests for Git URL sanitization.</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

pkg/utils/git/git_test.go

<ul><li>Adds 17 new test cases for <code>StripGitCredentials</code>.<br> <li> Adds 24 new test cases for <code>NormalizeGitURL</code>.<br> <li> Covers numerous URL formats, schemes, and edge cases.</ul>

</details>

  </td>
  <td><a href="https://github.com/snyk/go-application-framework/pull/586/changes#diff-8232e3bc8efc6284c4d55255fbef42ed5b93865b132e782d958a6bceaa7d9276">+71/-0</a>&nbsp; &nbsp; </td>

</tr>
</table></td></tr></tr></tbody></table>

</details>

___

### CODE REVIEW

File: pkg/utils/git/git.go [hunk 1]
```diff
@@ -54,6 +56,118 @@ func RepoFromDir(inputDir string) (*git.Repository, *config.RemoteConfig, error)
...
+	}
+
+	// Handle SCP-style on the original input before credential stripping.
+	// StripGitCredentials turns user@host:path into host:path, which url.Parse
+	// would misinterpret as scheme:path. So we extract host and path directly.
+	if isSCPStyle(rawUrl) {
+		atIdx := strings.Index(rawUrl, "@")
+		afterAt := rawUrl[atIdx+1:]
+		parts := strings.SplitN(afterAt, ":", 2)
+		if len(parts) == 2 {
```
Comment (rrama) [inline pkg/utils/git/git.go]: Do we need to ensure there is no `/` in `parts[0]`? Unsure if a `:` after the `/` is allowed or not in any protocols. If it does is then we aren't actually in an SCP style protocol.

File: pkg/utils/git/git.go [hunk 2]
```diff
@@ -54,6 +56,118 @@ func RepoFromDir(inputDir string) (*git.Repository, *config.RemoteConfig, error)
...
+
+	u, err := url.Parse(sanitized)
+	if err != nil {
+		return sanitized
+	}
+
+	switch u.Scheme {
+	case "ssh", "git+ssh", "ssh+git":
+		u.Scheme = "https"
+		u.User = nil
```
Comment (rrama) [inline pkg/utils/git/git.go]: Should we not be doing this for `http` and `https` as well? Should we not do it as a blanket set for all cases just in case?

File: pkg/utils/git/git.go [hunk 3]
```diff
@@ -54,6 +56,118 @@ func RepoFromDir(inputDir string) (*git.Repository, *config.RemoteConfig, error)
...
+		return u.String()
+	case "git", "http":
+		u.Scheme = "https"
+		u.Path = strings.TrimSuffix(u.Path, ".git")
+		return u.String()
+	case "https":
+		u.Path = strings.TrimSuffix(u.Path, ".git")
+		return u.String()
+	default:
+		return sanitized
```
Comment (rrama) [inline pkg/utils/git/git.go:173]: Is it not better to err on the side of caution, log that we don't support the schema and then return `""`?
We won't understand the protocol in the backend either, so may as well keep them all in sync and only allow protocols we know how to sanitize.
I'd suggest we do this fail-closed security approach elsewhere too.

File: pkg/utils/git/git.go [hunk 4]
```diff
@@ -54,6 +56,118 @@ func RepoFromDir(inputDir string) (*git.Repository, *config.RemoteConfig, error)
...
+	// StripGitCredentials turns user@host:path into host:path, which url.Parse
+	// would misinterpret as scheme:path. So we extract host and path directly.
+	if isSCPStyle(rawUrl) {
+		atIdx := strings.Index(rawUrl, "@")
+		afterAt := rawUrl[atIdx+1:]
+		parts := strings.SplitN(afterAt, ":", 2)
+		if len(parts) == 2 {
+			host := parts[0]
+			path := strings.TrimSuffix(parts[1], ".git")
+			return "https://" + host + "/" + path
```
Comment (rrama) [inline pkg/utils/git/git.go]: Should check for absolute path SCP URLs and not insert the `/` if it is.

File: pkg/utils/git/git.go [hunk 5]
```diff
@@ -54,6 +56,118 @@ func RepoFromDir(inputDir string) (*git.Repository, *config.RemoteConfig, error)
...
+	case "ssh", "git+ssh", "ssh+git":
+		u.Scheme = "https"
+		u.User = nil
+		// Strip SSH port — it's protocol-specific (e.g., 22, 7999) and doesn't map
+		// to the HTTPS port. HTTP/HTTPS ports are preserved in their respective cases
+		// because they represent the actual service port for that protocol.
+		u.Host = u.Hostname()
+		u.Path = strings.TrimSuffix(u.Path, ".git")
+		return u.String()
+	case "git", "http":
```
Comment (rrama) [inline pkg/utils/git/git.go]: Need to strip the port for the `git` protocol as well.

File: pkg/utils/git/git.go [hunk 6]
```diff
@@ -54,6 +56,118 @@ func RepoFromDir(inputDir string) (*git.Repository, *config.RemoteConfig, error)
...
+
+// stripCredentialsByPattern removes user[:pass]@ from a URL string using pattern matching.
+// Used as a fallback when url.Parse fails on malformed URLs that still contain credentials.
+func stripCredentialsByPattern(rawUrl string) string {
+	schemeEnd := strings.Index(rawUrl, "://")
+	if schemeEnd < 0 {
+		return rawUrl
+	}
+	rest := rawUrl[schemeEnd+3:]
+	atIdx := strings.Index(rest, "@")
```
Comment (rrama) [inline pkg/utils/git/git.go]: Passwords can have `@`s in them 🙃

File: pkg/utils/git/git.go [hunk 7]
```diff
@@ -54,6 +56,118 @@ func RepoFromDir(inputDir string) (*git.Repository, *config.RemoteConfig, error)
...
+	// so removing the username is safe and prevents token-as-username leakage.
+	if isSCPStyle(rawUrl) {
+		atIdx := strings.Index(rawUrl, "@")
+		return rawUrl[atIdx+1:]
+	}
+
+	u, err := url.Parse(rawUrl)
+	if err != nil {
+		// url.Parse failed (e.g., invalid port) — fall back to pattern-based stripping
+		return stripCredentialsByPattern(rawUrl)
```
Comment (rrama) [inline pkg/utils/git/git.go:92]: Or can we not just admit defeat, log a warning and return `""`?

File: pkg/utils/git/git.go (hunk 6 above)
Comment (bastiandoetsch) [reply pkg/utils/git/git.go]: True, but wouldn't that be URL encoded?

File: pkg/utils/git/git.go (hunk 1 above)
Comment (bastiandoetsch) [reply pkg/utils/git/git.go]: I don't think that's possible.

**Review by rrama — APPROVED** [review]
Discussed and decided against some of the changes I proposed. All good to merge now.

**Comment by bastiandoetsch** [comment]
Follow-up summary:

- Fixed SCP-style credential stripping to only treat `@` as userinfo when it appears before the first `:`.
- Prevents incorrect truncation when `@` appears in the repo path (for example: `github.com:org/repo@v2.git`).
- Added regression tests in `TestStripGitCredentials` and `TestNormalizeGitURL` for `@`-in-path scenarios.
- Verified with `go test ./pkg/utils/git -v` (pass).

**Comment by bastiandoetsch** [comment]
Addressed review follow-ups in this push:

- Fixed fallback credential stripping to handle @ inside malformed password/userinfo without leaking partial secrets (authority uses last @).
- Simplified NormalizeGitURL flow by reusing a single SCP-style detection result.
- Removed SCP branch fallback that could return non-normalized host:path output from normalization flow.
- Added support for git+http and git+https scheme normalization to https.
- Added table tests for these regression cases.
